File: 155.Min_Stack.c
typedef struct {
    int *arr;
    int *min_arr;
    int min_top;
    int maxSize;
    int top;
} MinStack;

/** initialize your data structure here. */
MinStack* minStackCreate(int maxSize) {
    MinStack *minStack = (MinStack *) malloc(sizeof(MinStack));
    minStack->maxSize = maxSize;
    minStack->arr = (int *) malloc(sizeof(int) * maxSize);
    minStack->min_arr = (int *) malloc(sizeof(int) * maxSize);
    minStack->top = -1;
    minStack->min_top = -1;
    return minStack;
}

void minStackPush(MinStack* obj, int x) {
    if (obj->min_top < 0 || obj->min_arr[obj->min_top] >= x) {
        obj->min_arr[++(obj->min_top)] = x;
    }
    obj->arr[++(obj->top)] = x;
}

void minStackPop(MinStack* obj) {
    
    if (obj->min_arr[obj->min_top] == obj->arr[obj->top]) {
        obj->min_top --;
    }
    obj->top --;
}

int minStackTop(MinStack* obj) {
    return obj->arr[obj->top];
}

int minStackGetMin(MinStack* obj) {
    return obj->min_arr[obj->min_top];
}

void minStackFree(MinStack* obj) {
    free(obj->arr);
    free(obj->min_arr);
    free(obj);
}

/**
 * Your MinStack struct will be instantiated and called as such:
 * struct MinStack* obj = minStackCreate(maxSize);
 * minStackPush(obj, x);
 * minStackPop(obj);
 * int param_3 = minStackTop(obj);
 * int param_4 = minStackGetMin(obj);
 * minStackFree(obj);
 */

File: 155.Min_Stack.py
class MinStack(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.l = []
        self.min_stack = []

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        self.l.append(x)
        if not len(self.min_stack) or x <= self.min_stack[-1]:
            self.min_stack.append(x)

    def pop(self):
        """
        :rtype: void
        """
        x = self.l.pop()
        if len(self.min_stack) and x == self.min_stack[-1]:
            self.min_stack.pop()
        

    def top(self):
        """
        :rtype: int
        """
        return self.l[-1]

    def getMin(self):
        """
        :rtype: int
        """
        return self.min_stack[-1] 


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
